header-gossip-network.md

```diff
@@ -0,0 +1,143 @@
+    historical_epochs: List[bytes32, max_length=MAX_HISTORICAL_EPOCHS],
+    current_epoch: EpochAccumulator,
+]
+```
+
+The algorithm for managing the accumulator is as follows.
+
+> TODO: provide a written spec too
+
+```python
+def update_accumulator(accumulator: Accumulator, new_block_header: BlockHeader) -> None:
+    # get the previous total difficulty
+    if len(accumulator.current_epoch) == 0:
+        # genesis
+        last_total_difficulty = 0
+    else:
+        last_total_difficulty = accumulator.current_epoch[-1].total_difficulty
+
+    # check if the epoch accumulator is full.
+    if len(accumulator.current_epoch) == EPOCH_SIZE:
+        # compute the final hash for this epoch
+        epoch_hash = hash_tree_root(accumulator.current_epoch)
+        # append the hash for this epoch to the list of historical epochs
+        accumulator.historical_epochs.append(epoch_hash)
+        # initialize a new empy epoch
+        accumulator.current_epoch = []
+
+    # construct the concise record for the new header and add it to the current epoch.
+    header_record = HeaderRecord(header.hash, last_total_difficulty + header.difficulty)
+    accumulator.current_epoch.append(header_record)
+```
+
+### Growth over time
+
+Each `HeaderRecord` is 64 bytes meaning the `EpochAccumulator` can range from `0 - 64 * EPOCH_SIZE` bytes (0-512KB) over the course of a single epoch.
+
+The `Accumulator.historical_epochs` grows by 32 bytes per epoch.
+
+At a 13 second block time we expect:
+
+- ~1 epochs per day
+- ~6 epochs per week
+- ~25 epochs per month
+- ~296 epochs per year
+
+Ignoring fluxuations in the size of `Accumulator.current_epoch` we should expect the size of the accumulator to grow at a rate of roughly 10kb per year.
+
+
+## Wire Protocol
+
+The Header Gossip Network is an overlay network on the [Discovery V5 Protocol](https://github.com/ethereum/devp2p/blob/master/discv5/discv5-theory.md).  The overlay network uses the PING/PONG/FINDNODES/FOUNDNODES/FINDCONTENT/FOUNDCONTENT/OFFER/ACCEPT messages from the [Portal Wire Protocol](./portal-wire-protocol.md).
+
+### Distance Function
+
+The Header Gossip Network uses the standard XOR distance function.
+
+### PING payload
+
+TODO: specify what a node should do prior to having acquired a copy of the accumulator
+
+```
+Ping.custom_payload := ssz_serialize(custom_data)
+custom_data         := Container(accumulator_root_hash: Bytes32, fork_id: Bytes32, head_hash: Bytes32, head_td: uint256)
+```
+
+### PONG payload
+
+TODO: specify what a node should do prior to having acquired a copy of the accumulator
+
+```
+Pong.custom_payload := ssz_serialize(custom_data)
+custom_data         := Container(accumulator_root_hash: Bytes32, fork_id: Bytes32, head_hash: Bytes32, head_td: uint256)
+```
+
+## Content Keys
+
+### Accumulator Snapshot
+
+> Note: The `content-id` for this content key is not used for any purpose.
+
+```
+content_key  := Container(content_type: uint8, accumulator_root_hash: Bytes32)
```

Comment:
Considering the two Containers are different, I think it would be better to change this into an SSZ `Union`.